Terraform Crash: panic: EvalApplyPre for ... called with nil Change

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. For example, see [this case](https://github.com/danieldreier/terraform-issue-reproductions/tree/master/25719). Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
❯ terraform --version               
Terraform v0.14.4
+ provider registry.terraform.io/hashicorp/aws v3.23.0
+ provider registry.terraform.io/hashicorp/kubernetes v1.13.3
+ provider registry.terraform.io/hashicorp/local v2.0.0
+ provider registry.terraform.io/hashicorp/null v3.0.0
+ provider registry.terraform.io/hashicorp/random v3.0.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
...
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "13.2.1"

  manage_aws_auth = false

  cluster_name    = "${module.<redacted>_envs.aws_accounts[terraform.workspace]["account-name"]}-eks-cluster"
  cluster_version = "1.17"
  vpc_id          = module.vpc.vpc_id
  subnets         = module.vpc.private_subnets

  ## https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html
  enable_irsa = true

  workers_group_defaults = {
    target_group_arns = concat(
      module.eks_public_alb.target_group_arns,
      module.eks_private_alb.target_group_arns
    )

    tags = [
      {
        "key"                 = "k8s.io/cluster-autoscaler/${module.<redacted>_envs.aws_accounts[terraform.workspace]["account-name"]}-eks-cluster"
        "value"               = "owned"
        "propagate_at_launch" = true
      },
      {
        "key"                 = "k8s.io/cluster-autoscaler/enabled"
        "value"               = "true"
        "propagate_at_launch" = true
      }
    ]
  }

  worker_additional_security_group_ids = [module.eks_nodes_custom_security_group.this_security_group_id]

  worker_groups = [
    for private_subnet in module.vpc.private_subnets : {
      name          = "${module.<redacted>_envs.aws_accounts[terraform.workspace]["account-name"]}-eks-worker-ondemand-${private_subnet}"
      instance_type = module.<redacted>_envs.aws_accounts[terraform.workspace]["eks"]["ondemand_instance_type"]

      subnets = list(private_subnet)

      asg_max_size         = module.<redacted>_envs.aws_accounts[terraform.workspace]["eks"]["ondemand_max_size"]
      asg_min_size         = module.<redacted>_envs.aws_accounts[terraform.workspace]["eks"]["ondemand_min_size"]
      asg_desired_capacity = module.<redacted>_envs.aws_accounts[terraform.workspace]["eks"]["ondemand_desired_capacity"]

      kubelete_extra_args = "--node-labels=kubernetes.io/lifecycle=normal"
      public_ip           = false
    }
  ]

  tags = local.common_tags
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.main_eks.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.main_eks.certificate_authority.0.data)
  token                  = data.aws_eks_cluster_auth.main_eks.token
  load_config_file       = false
}

data "aws_eks_cluster" "main_eks" {
  name = module.eks.cluster_id
}

data "aws_eks_cluster_auth" "main_eks" {
  name = module.eks.cluster_id
}

...
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->
https://gist.github.com/lebenitza/9eb282bdec48e3d3652114d78ce85db4

### Expected Behavior
<!--
What should have happened?
-->
1. New terraform environment `terraform apply`
2. Part of the EKS module apply failed because of low AWS services quotas (1 autoscaling group from 3)
3. Increase the AWS service quota for EC2
4. `terraform apply` again or `terraform apply -target=module.eks` to delete and create the now defunct resource of the 3rd autoscaling group and have a consistent terraform state at the end

### Actual Behavior
<!--
What actually happened?
-->
5. Terraform crash
6. Tried different small version changes to AWS provider without success

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
See: Expected Behavior

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
Nothing fancy, local run of terraform CLI, remote S3/DynamoDB state with lock.

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->
Might be related to #27474

Forgot to mention: The resource (the autoscaling group) is created with no issues. In addition the defunct resource in the previous crash is never removed from AWS. They keep piling up if multiple applies are ran.

```terraform
module.eks.random_pet.workers[2]: Destroying... [id=ample-beagle]
module.eks.random_pet.workers[0]: Destroying... [id=evolving-goblin]
module.eks.random_pet.workers[1]: Destroying... [id=sharp-spider]
module.eks.random_pet.workers[1]: Destruction complete after 0s
module.eks.random_pet.workers[0]: Destruction complete after 0s
module.eks.random_pet.workers[2]: Destruction complete after 0s
module.eks.aws_launch_configuration.workers[0]: Destroying... [id=<redacted>-prod-eks-cluster-<redacted>-prod-eks-worker-ondemand-subnet-0304af192f1fa89642021011107332374880000000f]
module.eks.aws_launch_configuration.workers[2]: Destroying... [id=<redacted>-prod-eks-cluster-<redacted>-prod-eks-worker-ondemand-subnet-048a19a4e5b0f29a52021011107332373050000000d]
module.eks.aws_launch_configuration.workers[1]: Destroying... [id=<redacted>-prod-eks-cluster-<redacted>-prod-eks-worker-ondemand-subnet-0d8ad43734233bd082021011107332374570000000e]
module.eks.aws_autoscaling_group.workers[1]: Creating...
module.eks.aws_autoscaling_group.workers[1]: Still creating... [10s elapsed]
module.eks.aws_autoscaling_group.workers[1]: Still creating... [20s elapsed]
module.eks.aws_autoscaling_group.workers[1]: Still creating... [30s elapsed]
module.eks.aws_autoscaling_group.workers[1]: Still creating... [40s elapsed]
module.eks.aws_autoscaling_group.workers[1]: Creation complete after 46s [id=<redacted>-prod-eks-cluster-<redacted>-prod-eks-worker-ondemand-subnet-0d8ad43734233bd0820210112123151566300000001]
panic: EvalApplyPre for module.eks.aws_autoscaling_group.workers[1] called with nil Change
```

The crash happens exactly after the resource is created successfully.

--- a/terraform/node_resource_destroy_deposed.go
+++ b/terraform/node_resource_destroy_deposed.go
@@ -159,6 +159,11 @@ func (n *NodeDestroyDeposedResourceInstanceObject) Execute(ctx EvalContext, op w
 		return diags.Append(err)
 	}
 
+	if state == nil {
+		diags = diags.Append(fmt.Errorf("missing deposed state for %s (%s)", n.Addr, n.DeposedKey))
+		return diags
+	}
+
 	change, destroyPlanDiags := n.planDestroy(ctx, state, n.DeposedKey)
 	diags = diags.Append(destroyPlanDiags)
 	if diags.HasErrors() {

--- a/terraform/node_resource_destroy_deposed_test.go
+++ b/terraform/node_resource_destroy_deposed_test.go
@@ -176,3 +176,28 @@ aws_instance.foo: (1 deposed)
   Deposed ID 1 = i-abc123
 	`)
 }
+
+func TestNodeDestroyDeposedResourceInstanceObject_ExecuteMissingState(t *testing.T) {
+	p := simpleMockProvider()
+	ctx := &MockEvalContext{
+		StateState:           states.NewState().SyncWrapper(),
+		ProviderProvider:     simpleMockProvider(),
+		ProviderSchemaSchema: p.ProviderSchema(),
+		ChangesChanges:       plans.NewChanges().SyncWrapper(),
+	}
+
+	node := NodeDestroyDeposedResourceInstanceObject{
+		NodeAbstractResourceInstance: &NodeAbstractResourceInstance{
+			Addr: mustResourceInstanceAddr("test_object.foo"),
+			NodeAbstractResource: NodeAbstractResource{
+				ResolvedProvider: mustProviderConfig(`provider["registry.terraform.io/hashicorp/test"]`),
+			},
+		},
+		DeposedKey: states.NewDeposedKey(),
+	}
+	err := node.Execute(ctx, walkApply)
+
+	if err == nil {
+		t.Fatal("expected error")
+	}
+}

